### ISSUE TITLE
Match gather implementation to torch

### ISSUE THREAD

**Issue by dosier** [tinygrad/tinygrad#1351]
I tried replacing the one used in the vits example with the newly merged one (#1329 ) but it's giving errors, maybe for sake of ease of porting models we should make sure the input/output results match that of torch.

**Comment by geohotstan** [comment]
Do you have code I can run to reproduce the error?

**Comment by dosier** [comment]
> Do you have code I can run to reproduce the error?
Yes, sorry. 
```py
from tinygrad.tensor import Tensor
import torch, numpy

tiny_idx = Tensor.ones(361, 1)
tiny_cum_widths = Tensor.randn(361, 11)
tiny_res = tiny_cum_widths.gather(tiny_idx, dim=-1).numpy()

torch_idx = torch.ones(361, 1, dtype=torch.long)
torch_cum_widths = torch.randn(361, 11)
torch_res = torch_cum_widths.gather(-1, torch_idx).numpy()

print(f"tiny_shape: {tiny_res.shape}")
print(f"torch_shape: {torch_res.shape}")
assert numpy.allclose(torch_res, tiny_res)
```

Which prints:
```
tiny_shape: (11, 361, 1, 361)
torch_shape: (361, 1)
```

**Comment by geohotstan** [comment]
<img width="822" alt="image" src="https://github.com/tinygrad/tinygrad/assets/135171913/1d93deb8-92dd-43fc-8844-301ce8fbcac6">

Essentially what `Tensor.gather()` does right now is
```
out[i][j][k] = input[index[i][j][k]][:][:]  # if dim == 0
out[i][j][k] = input[:][index[i][j][k]][:]  # if dim == 1
out[i][j][k] = input[:][:][index[i][j][k]]  # if dim == 2
```

the `input[index[i][j][k]][j][k]` in `torch.gather` requires the ability to do combined indexing https://github.com/tinygrad/tinygrad/issues/967 (of which I have yet to figure out haha! :D)

There are other ways to hack `torch.gather()` aka GatherElements https://github.com/geohotstan/tinygrad/blob/passing_onnx/extra/onnx_ops.py#L439, but that's probably not ideal either.

I'll try at combined/fancy indexing some more, and if I figure it out, I'll move the current `Tensor.gather()` into `Tensor.__getitem__()` and express the new `Tensor.gather()` using `Tensor.__getitem__()`?
kinda like the torch example.

this way the apis can match

**Comment by geohotstan** [comment]
ok I think I've got a working `torch.gather()` based on the logic of combined indexing (dim by dim reduction)
```
def torch_gather(self: Tensor, idx: Tensor, dim):
  if dim < 0: dim += self.ndim
  ret = self.gather(idx, dim=dim)
  self_extra = self.shape[:dim] + self.shape[dim+1:]
  idx_extra = idx.shape[:dim] + idx.shape[dim+1:]
  idx_extra = [[dim+n,i] if n < dim else (dim+1+n, i) for n,i in enumerate(idx_extra)][::-1]
  self_extra = [(n,i) if n < dim else(n+idx.ndim, i) for n,i in enumerate(self_extra)][::-1]
  for n, ((dim_idx, idx), (dim_self, self)) in enumerate(zip(idx_extra, self_extra)):
    if dim_self < dim_idx and n < len(idx_extra)-1: idx_extra[n+1][0] -= 1
    arange_idx = Tensor.arange(idx).reshape(*[1]*dim_idx, idx, *[1]*(ret.ndim-dim_idx-1))
    arange_self = Tensor.arange(self).reshape(*[1]*dim_self, self, *[1]*(ret.ndim-dim_self-1))
    ret = ((arange_idx == arange_self) * ret).sum(dim_self)

  torch_ans = torch.tensor(self.numpy()).gather(index=torch.tensor(idx.numpy().astype(np.int64)), dim=dim)
  np.testing.assert_allclose(ret.numpy(), torch_ans.numpy())
  # print(ret.numpy())
  # print(torch_ans)

tiny_idx = Tensor.ones(361, 1)
tiny_cum_widths = Tensor.randn(361, 11)
torch_gather(tiny_cum_widths, tiny_idx, -1)
```
code looks kinda disgusting though... but it works for all onnx GatherElements tests :D 
Also not sure about performance compared to hacked GatherElements...

but I believe this logic can be extrapolated to work for fancy indexing as well.
